pallets/nomad/home/src/lib.rs

```diff
@@ -0,0 +1,363 @@
+			}
+		}
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+		fn build(&self) {
+			<Base<T>>::put(NomadBase::new(
+				self.local_domain,
+				self.committed_root,
+				self.updater,
+			));
+			<Tree<T>>::put(NomadLightMerkle::default());
+		}
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		Dispatch {
+			message_hash: H256,
+			leaf_index: u32,
+			destination_and_nonce: u64,
+			committed_root: H256,
+			message: Vec<u8>,
+		},
+		Update {
+			home_domain: u32,
+			previous_root: H256,
+			new_root: H256,
+			signature: Vec<u8>,
+		},
+		ImproperUpdate {
+			previous_root: H256,
+			new_root: H256,
+			signature: Vec<u8>,
+		},
+		UpdaterSlashed {
+			updater: H160,
+			reporter: T::AccountId,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		InitializationError,
+		IngestionError,
+		SignatureRecoveryError,
+		MessageTooLarge,
+		InvalidUpdaterSignature,
+		CommittedRootNotMatchUpdatePrevious,
+		RootForIndexNotFound,
+		IndexForRootNotFound,
+		FailedState,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T>
+	where
+		[u8; 32]: From<T::AccountId>,
+	{
+		/// Dispatch a message to the destination domain and recipient address.
+		#[pallet::weight(100)]
+		pub fn dispatch(
+			origin: OriginFor<T>,
+			destination_domain: u32,
+			recipient_address: H256,
+			message_body: BoundedVec<u8, T::MaxMessageBodyBytes>,
+		) -> DispatchResult {
+			let sender: [u8; 32] = ensure_signed(origin)?.into();
+			Self::do_dispatch(
+				sender.into(),
+				destination_domain,
+				recipient_address,
+				message_body,
+			)
+		}
+
+		/// Verify/submit signed update.
+		#[pallet::weight(100)]
+		pub fn update(origin: OriginFor<T>, signed_update: SignedUpdate) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+			Self::do_update(sender, signed_update)
+		}
+
+		/// Verify/slash updater for improper update.
+		#[pallet::weight(100)]
+		pub fn improper_update(
+			origin: OriginFor<T>,
+			signed_update: SignedUpdate,
+		) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+			Self::do_improper_update(sender, &signed_update)?;
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T>
+	where
+		[u8; 32]: From<T::AccountId>,
+	{
+		fn ensure_not_failed() -> Result<(), Error<T>> {
+			ensure!(
+				Self::base().state != NomadState::Failed,
+				Error::<T>::FailedState
+			);
+			Ok(())
+		}
+
+		/// Format message, insert hash into merkle tree, and update mappings
+		/// between tree roots and message indices.
+		pub fn do_dispatch(
+			sender: H256,
+			destination_domain: u32,
+			recipient_address: H256,
+			message_body: BoundedVec<u8, T::MaxMessageBodyBytes>,
+		) -> DispatchResult {
+			Self::ensure_not_failed()?;
+
+			// Check message length against max
+			let message_length = message_body.len() as u32;
+			ensure!(
+				message_length < T::MaxMessageBodyBytes::get(),
+				Error::<T>::MessageTooLarge
+			);
```

Comment:
Is not already checked by the type itself (as `BoundedVec<u8, T::MaxMessageBodyBytes>`)?
```suggestion
```